src/drivers/virtio/virtqueue/mod.rs

```diff
@@ -512,27 +512,65 @@ pub enum BufferType {
 	Indirect,
 }
 
-/// MemPool allows to easily control, request and provide memory for Virtqueues.
-///
-/// The struct is initialized with a limit of free running "tracked"
-/// memory descriptor ids. As Virtqueus do only allow a limited amount of descriptors in their queue,
-/// the independent queues, can control the number of descriptors by this.
-struct MemPool {
-	pool: Vec<u16>,
-	limit: u16,
-}
+mod index_alloc {
+	use alloc::boxed::Box;
+
+	use align_address::Align;
 
-impl MemPool {
-	/// Returns a given id to the id pool
-	fn ret_id(&mut self, id: u16) {
-		self.pool.push(id);
+	/// This type allows allocating indices.
+	///
+	/// The indices can be used as descriptor IDs.
+	pub struct IndexAlloc {
+		/// Zero bits are available.
+		bits: Box<[usize]>,
 	}
 
-	/// Returns a new instance, with a pool of the specified size.
-	fn new(size: u16) -> MemPool {
-		MemPool {
-			pool: (0..size).collect(),
-			limit: size,
+	const USIZE_BITS: usize = usize::BITS as usize;
+
+	impl IndexAlloc {
+		pub fn new(len: usize) -> Self {
+			let usizes = len.align_up(USIZE_BITS) / USIZE_BITS;
+			let extra_bits = len % USIZE_BITS;
+
+			let mut bits = vec![0; usizes].into_boxed_slice();
+
+			if extra_bits != 0 {
+				*bits.last_mut().unwrap() |= usize::MAX >> extra_bits;
```

Comment:
```suggestion
				*bits.last_mut().unwrap() = usize::MAX >> extra_bits;
```
Should not change the result but I think it makes the intention clearer.